File: recruitpro-api/app/services/designation_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException, status

from app.models.department import Department
from app.repositories.designation_repository import DesignationRepository
from app.schemas.designation_schema import (
    DesignationCreate,
    DesignationUpdate,
)


class DesignationService:

    def __init__(self):
        self.repository = DesignationRepository()

    # -------------------------
    # Get All Designations
    # -------------------------
    def get_all_designations(
        self,
        db: Session,
        current_user,
        search: str = "",
        company_id: int = None,
        department_id: int = None,
        sort_by: str = "DesignationName",
        order: str = "asc",
        page: int = 1,
        page_size: int = 10,
    ):

        return self.repository.get_all(
            db,
            current_user,
            search,
            company_id,
            department_id,
            sort_by,
            order,
            page,
            page_size,
        )

    # -------------------------
    # Get Designation By Id
    # -------------------------
    def get_designation_by_id(
        self,
        db: Session,
        designation_id: int,
        current_user,
    ):

        designation = self.repository.get_by_id(
            db,
            designation_id,
        )

        if not designation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Designation not found",
            )

        # Company users can access only their own company
        if (
            not current_user["is_super_admin"]
            and designation.CompanyId != current_user["company_id"]
        ):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You are not authorized to access this designation",
            )

        return designation

    # -------------------------
    # Create Designation
    # -------------------------
    def create_designation(
        self,
        db: Session,
        designation: DesignationCreate,
        current_user,
    ):

        # Company users can create only in their own company
        if not current_user["is_super_admin"]:
            designation.CompanyId = current_user["company_id"]

        # Check Department Exists
        department = (
            db.query(Department)
            .filter(Department.DepartmentId == designation.DepartmentId)
            .first()
        )

        if not department:
            raise HTTPException(
                status_code=400,
                detail="Department not found",
            )

        # Check Department belongs to Company
        if department.CompanyId != designation.CompanyId:
            raise HTTPException(
                status_code=400,
                detail="Selected Department does not belong to the selected company",
            )

        # Duplicate Code
        existing_code = self.repository.get_by_code(
            db,
            designation.DesignationCode,
            designation.CompanyId,
        )

        if existing_code:
            raise HTTPException(
                status_code=400,
                detail="Designation Code already exists in this company",
            )

        # Duplicate Name
        existing_name = self.repository.get_by_name(
            db,
            designation.DesignationName,
            designation.CompanyId,
        )

        if existing_name:
            raise HTTPException(
                status_code=400,
                detail="Designation Name already exists in this company",
            )

        return self.repository.create(
            db,
            designation,
        )

    # -------------------------
    # Update Designation
    # -------------------------
    def update_designation(
        self,
        db: Session,
        designation_id: int,
        designation: DesignationUpdate,
        current_user,
    ):

        existing_designation = self.repository.get_by_id(
            db,
            designation_id,
        )

        if not existing_designation:
            raise HTTPException(
                status_code=404,
                detail="Designation not found",
            )

        # Company users can update only their own company
        if (
            not current_user["is_super_admin"]
            and existing_designation.CompanyId != current_user["company_id"]
        ):
            raise HTTPException(
                status_code=403,
                detail="You are not authorized to update this designation",
            )

        if not current_user["is_super_admin"]:
            designation.CompanyId = current_user["company_id"]

        # Check Department Exists
        department = (
            db.query(Department)
            .filter(Department.DepartmentId == designation.DepartmentId)
            .first()
        )

        if not department:
            raise HTTPException(
                status_code=400,
                detail="Department not found",
            )

        # Check Department belongs to Company
        if department.CompanyId != designation.CompanyId:
            raise HTTPException(
                status_code=400,
                detail="Selected Department does not belong to the selected company",
            )

        # Duplicate Code
        duplicate_code = self.repository.get_by_code(
            db,
            designation.DesignationCode,
            designation.CompanyId,
        )

        if duplicate_code and duplicate_code.DesignationId != designation_id:
            raise HTTPException(
                status_code=400,
                detail="Designation Code already exists in this company",
            )

        # Duplicate Name
        duplicate_name = self.repository.get_by_name(
            db,
            designation.DesignationName,
            designation.CompanyId,
        )

        if duplicate_name and duplicate_name.DesignationId != designation_id:
            raise HTTPException(
                status_code=400,
                detail="Designation Name already exists in this company",
            )

        return self.repository.update(
            db,
            designation_id,
            designation,
        )

    # -------------------------
    # Delete Designation
    # -------------------------
    def delete_designation(
        self,
        db: Session,
        designation_id: int,
        current_user,
    ):

        designation = self.repository.get_by_id(
            db,
            designation_id,
        )

        if not designation:
            raise HTTPException(
                status_code=404,
                detail="Designation not found",
            )

        # Company users can delete only their own company
        if (
            not current_user["is_super_admin"]
            and designation.CompanyId != current_user["company_id"]
        ):
            raise HTTPException(
                status_code=403,
                detail="You are not authorized to delete this designation",
            )

        self.repository.delete(
            db,
            designation_id,
        )

        return {"message": "Designation deleted successfully"}

File: recruitpro-ui/src/app/services/designation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class DesignationService {

  private apiUrl = 'http://127.0.0.1:8000/designations';

  constructor(private http: HttpClient) { }

  // ==========================
  // Get All Designations
  // ==========================
  getDesignations(
    search: string = '',
    companyId: number | null = null,
    departmentId: number | null = null,
    sortBy: string = 'DesignationName',
    order: string = 'asc',
    page: number = 1,
    pageSize: number = 10
  ): Observable<any> {

    let params = new HttpParams()
      .set('search', search)
      .set('sort_by', sortBy)
      .set('order', order)
      .set('page', page)
      .set('page_size', pageSize);

    if (companyId !== null) {
      params = params.set('company_id', companyId);
    }

    if (departmentId !== null) {
      params = params.set('department_id', departmentId);
    }

    return this.http.get<any>(`${this.apiUrl}/`, { params });

  }

  // ==========================
  // Get Designation By Id
  // ==========================
  getDesignationById(id: number): Observable<any> {

    return this.http.get<any>(`${this.apiUrl}/${id}`);

  }

  // ==========================
  // Add Designation
  // ==========================
  addDesignation(data: any): Observable<any> {

    return this.http.post<any>(`${this.apiUrl}/`, data);

  }

  // ==========================
  // Update Designation
  // ==========================
  updateDesignation(
    id: number,
    data: any
  ): Observable<any> {

    return this.http.put<any>(
      `${this.apiUrl}/${id}`,
      data
    );

  }

  // ==========================
  // Delete Designation
  // ==========================
  deleteDesignation(id: number): Observable<any> {

    return this.http.delete<any>(
      `${this.apiUrl}/${id}`
    );

  }

}